tensorflow_federated/python/common_libs/execution_tracing.py

```diff
@@ -0,0 +1,266 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tracing provider that creates human-readable logging for computations.
+
+This module is helpful for understanding how computations are handled by an
+established set of executor stacks. This process can be opaque if a user is
+unfamiliar with a particular executor handles or modifies a specific
+Computation as it makes its way through the executor stack. This
+TracingProvider logs information about the computation in a presentable manner.
+Note that it does not attempt to correlate asynchronous calls, and so using
+this in conjunction with a ThreadDelegatingExecutor will not preserve logging
+order as it otherwise would.
+"""
+from enum import Enum
+import logging
+import re
+from typing import Any, Dict, Optional, Tuple
+
+from tensorflow_federated.python.common_libs import tracing
+
+
+class TraceFilterMode(Enum):
+  VALUES = 0
+  INTRINSICS = 1
+  ALL = 2
+
+
+def _pattern_union(*patterns):
+  pattern_strs = map(lambda p: p.pattern, patterns)
+  return re.compile('|'.join(pattern_strs))
+
+
+_VALUE_PATTERN = re.compile('^create_([a-z]+)$')
+_INTRINSIC_PATTERN = re.compile('^_compute_intrinsic_([a-z_]+)$')
+
+_TRACE_FILTER_PATTERNS = {
+    TraceFilterMode.VALUES:
+        _VALUE_PATTERN,
+    TraceFilterMode.INTRINSICS:
+        _pattern_union(_VALUE_PATTERN, _INTRINSIC_PATTERN),
+    TraceFilterMode.ALL:
+        None,
+}
+
+DEFAULT_FORMAT_STRATEGY = lambda x: '{} {}'.format(x, type(x))
+
+
+class ExecutionTracingProvider(tracing.TracingProvider):
+  """Traces a subset of functions appearing in the current execution context.
+
+  Note: This TracingProvider does not attempt to correlate log messages across
+  threads, so executor stacks using ThreadDelegatingExecutor will generally lose
+  their nesting structure in the logs.
+
+  This implementation does not require storing additional trace context state,
+  so most methods are no-ops.
+  """
+
+  def __init__(self,
+               trace_filter_mode=TraceFilterMode.INTRINSICS,
+               default_format_strategy=DEFAULT_FORMAT_STRATEGY):
+    assert isinstance(trace_filter_mode, TraceFilterMode)
+
+    # establish what we're tracing
+    self._is_traceable_regex = _TRACE_FILTER_PATTERNS[trace_filter_mode]
+
+    # fill in formatting strategies
+    self._default_strategy = default_format_strategy
+    self._formatting_strategies = dict()
+    self._make_formatting_strategies()
+
+    # indentation management
+    # start at -1 so the first call prints at 0
+    self._current_call_level = -1
```

Comment:
I was kind of hoping that the indentation level here could be moved into the span yield infra. That is, that the output data could be logically organized underneath its parent by using the parenting abilities offered by `span`. This would mean that either (a) output wouldn't be available to the end of the program or (b) the output would be printed sort of "backwards", listing parent spans *below* their children. Both of these are sort of unfortunate, but either would make this provider compatible with the more complicated (but more common) executor setups involving concurrency. What do you think about this idea? I feel somewhat uneasy about offering a debugging utility that will print incorrect results for most users' default configurations.

Reply:
I like this idea a lot, I hadn't quite connected those dots before. When I get a chance to revisit this PR, I'll give this a shot